enhancements/authentication/custom-scc-preemption-prevention.md

```diff
@@ -0,0 +1,252 @@
+
+## Proposal
+
+### "hard protection" - Requesting a required SCC
+
+This proposal suggests to implement a mechanism which enables OpenShift core workloads to enforce
+an SCC to be pinned to a workload.
+This is accomplished by setting an annotation named `openshift.io/required-scc`
+whose value contains the name of the SCC that is supposed to be pinned.
+
+If the requested SCC does not exist in the system, admission fails with an error.
+
+If the requested SCC does exist, any other SCC available in the system is ignored
+and the requested SCC is the only one being considered for admission.
+
+If the SCC is applicable to the workload, it is being used,
+else admission fails with an error.
+
+Pros:
+- The required and pinned SCC is deterministic.
+- No other SCC in the system is taken into account.
+
+Cons:
+- If the required SCC is not applicable, the workload fails to be scheduled.
+This can be detected in the existing e2e suite.
+- All core OpenShift workloads need to set that annotation in their deployment artifacts.
+This implies that, after the change merges, many repositories have to be touched.
+
+An existing implementation is available at https://github.com/openshift/apiserver-library-go/pull/90.
+
+### (Alternative) "soft protection" - Hinting a favored SCC
+
+Another suggested approach is to use hinting instead.
+Here, an annotation named `openshift.io/scc-hint` contains the hinted SCC name as its value.
+
+If the hinted SCC does not exist in the system, it is being ignored silently.
+In this case SCC admission continues with all remaining SCCs available in the system as described above.
+
+If the hinted SCC does exist in the system it is being prepended to the sorted list describe above,
+before the highest priority SCC.
+
+This ensures that the hinted SCC is deterministically considered to be applied first
+before any other SCC available in the system.
+
+Pros:
+- Less hard failure modes than the proposed "hard protection" mechanism.
+- Less intrusive than the "hard protection" alternative
+as the workload will continue to be asserted against the remaining SCCs available in the system.
+
+Cons:
+- If the hinted SCC is not applicable to the workload,
+admission continues with asserting all remaining SCCs available in the system as described above.
+- If the hinted SCC is not applicable, admission continues with asserting all the other SCCs available in the system,
+potentially even custom SCCs which still can cause problems.
+- All core OpenShift workloads need to set that annotation in their deployment artifacts.
+This implies that, after the change merges, many repositories have to be touched.
+- To verify that hinted SCCs are really being applied, an e2e test has to be written.
+This can be done centrally.
+
+An existing implementation is available at https://github.com/openshift/apiserver-library-go/pull/105.
+
+This proposal favors the "hard protection" variant and continues to describe that one further on.
```

Comment:
In case of updates, pod spec mutation must not be applied as per current logic. Admission then behaves similar to validation. If I am not allowed to use the requested SCC, a validation error will be returned with the reason.

Reply:
Added commit which clarifies the current behavior, PTAL. Do you think we'll have to do bigger changes to the logic here?